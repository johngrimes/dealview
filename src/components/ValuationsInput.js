import React from 'react'
import DatePicker from 'react-datepicker'
import moment from 'moment'
import _ from 'lodash'

import HiddenField from './HiddenField.js'
import { DateDisplayFormat, DateStorageFormat } from '../data/commonTypes.js'
import { ValuationDefault, compareValuationsByDate } from '../data/valuations.js'
import * as Validations from '../utils/formValidation.js'

import './styles/ValuationsInput.css'
import 'react-datepicker/dist/react-datepicker-cssmodules.css'

class ValuationsInput extends React.Component {
  constructor(props) {
    super(props)
    const valuations = this.props.valuations
      ? this.props.valuations.sort(compareValuationsByDate)
      : this.props.valuations
    this.state = {
      valuations,
      touched: false,
    }
    this._refs = {}

    this.handleFocus = this.handleFocus.bind(this)
    this.handleBlur = this.handleBlur.bind(this)
    this.handleAddValuation = this.handleAddValuation.bind(this)
  }

  handleChange(i, field, value) {
    if (this.state.valuations === undefined) return
    this.setState(
      prevState => {
        const updatedValuations = prevState.valuations.slice()
        updatedValuations[i] = { ...updatedValuations[i], [field]: value }
        return {
          ...prevState,
          valuations: updatedValuations,
          touched: true,
        }
      },
      () => {
        if (this.state.valuations && this.props.onChange) {
          this.props.onChange(this.state.valuations)
        }
      }
    )
  }

  handleAddValuation() {
    const updatedValuations = this.state.valuations ? this.state.valuations.concat([ValuationDefault]) : []
    this.setState(
      () => ({ valuations: updatedValuations }),
      () => {
        if (this.props.onChange) this.props.onChange(updatedValuations)
      }
    )
  }

  handleDeleteValuation(index) {
    this.setState(
      prevState => ({
        valuations: prevState.valuations.filter((_, i) => i !== index),
      }),
      () => {
        if (this.props.onChange && this.state.valuations) {
          this.props.onChange(this.state.valuations)
        }
      }
    )
  }

  handleFocus(event) {
    const target = event.target
    if (this.props.onFocus) {
      this.props.onFocus(target.name)
    }
  }

  setFocus() {
    if (this.props.focus && this._refs[this.props.focus]) {
      this._refs[this.props.focus].focus()
    }
  }

  handleBlur(event) {
    if (event.relatedTarget && !event.relatedTarget.name.match(`^${this.props.name}`)) {
      this.setState(prevState => ({
        valuations: prevState.valuations.sort(compareValuationsByDate),
      }))
    }
  }

  componentWillReceiveProps(props) {
    if (props.valuations) {
      const sortedValuations = props.valuations.sort(compareValuationsByDate)
      this.setState(() => ({ valuations: sortedValuations }))
    }
  }

  componentDidMount() {
    this.setFocus()
  }

  componentDidUpdate() {
    this.setFocus()
  }

  render() {
    const { errors, forceErrorDisplay, minDate, maxDate } = this.props
    const { touched } = this.state
    const valuations =
      this.state.valuations === undefined
        ? []
        : this.state.valuations.map((v, i) => {
          return (
              <tr key={i} className={v.type && v.type !== 'none' ? `valuation valuation-${v.type}` : 'valuation'}>
                <td className='valuations-date'>
                  <DatePicker
                    name={`valuations-date-${i}`}
                    dateFormat={DateDisplayFormat}
                    selected={v.date ? moment(v.date, DateStorageFormat) : undefined}
                    showYearDropdown
                    minDate={v.type !== 'purchase' ? minDate : undefined}
                    maxDate={v.type !== 'sale' ? maxDate : undefined}
                    autoFocus={this.props.focus === `valuations-date-${i}`}
                    onChange={moment => this.handleChange(i, 'date', moment.format(DateStorageFormat))}
                    onFocus={this.handleFocus}
                    onBlur={this.handleBlur}
                  />
                </td>
                <td className='valuations-amount'>
                  <input
                    name={`valuations-amount-${i}`}
                    type='number'
                    min='1'
                    value={v.amount ? v.amount.toString() : ''}
                    placeholder='Value'
                    onChange={event => this.handleChange(i, 'amount', parseInt(event.target.value, 10))}
                    ref={input => (this._refs[`valuations-amount-${i}`] = input)}
                    onFocus={this.handleFocus}
                    onBlur={this.handleBlur}
                  />
                </td>
                <td className='valuations-note'>
                  <input
                    name={`valuations-note-${i}`}
                    type='text'
                    value={v.note}
                    placeholder='Note'
                    onChange={event => this.handleChange(i, 'note', event.target.value)}
                    ref={input => (this._refs[`valuations-note-${i}`] = input)}
                    onFocus={this.handleFocus}
                    onBlur={this.handleBlur}
                  />
                  <HiddenField name={`valuations-type-${i}`} type='hidden' value={v.type} />
                </td>
                <td className='valuations-delete'>
                  <button
                    className='delete-valuation-button'
                    type='button'
                    onClick={() => this.handleDeleteValuation(i)}
                  />
                </td>
              </tr>
          )
        })
    const errorTags = []
    if (errors) {
      errors.forEach((msg, i) =>
        errorTags.push(
          <div key={i} className='error'>
            {msg}
          </div>
        )
      )
    }

    return (
      <div className='valuations-input control-group'>
        <table className='valuations table'>
          <tbody>
            {valuations}
          </tbody>
        </table>
        {(forceErrorDisplay || touched) &&
          errorTags.length > 0 &&
          <div className='errors'>
            {errorTags}
          </div>}
        <button className='button add-valuation-button' type='button' onClick={this.handleAddValuation}>
          +&nbsp;&nbsp;Add
        </button>
      </div>
    )
  }

  static validate(valuations) {
    return []
      .concat(ValuationsInput.valuationsHaveDates(valuations))
      .concat(ValuationsInput.valuationsHaveAmounts(valuations))
      .concat(ValuationsInput.valuationsDontShareDates(valuations))
  }

  static valuationsHaveDates(valuations) {
    return Validations.validate(valuations, vals => vals.every(v => v.date), 'Valuations must all have dates')
  }

  static valuationsHaveAmounts(valuations) {
    return Validations.validate(valuations, vals => vals.every(v => v.amount), 'Valuations must all have amounts')
  }

  static valuationsDontShareDates(valuations) {
    return Validations.validate(
      valuations,
      vals => {
        const valsWithDates = vals.filter(v => v.date)
        return valsWithDates.length === _.uniq(valsWithDates.map(v => v.date)).length
      },
      'Valuations must all be on different dates'
    )
  }
}

export default ValuationsInput
