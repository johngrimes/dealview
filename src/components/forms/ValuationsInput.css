.add-valuation-button {
  margin-top: 8px;
}

.valuations-input table .control-group input {
  border: 0;
  box-shadow: none;
  width: 100px;
  padding: 8px 16px;
}

.valuations-input table input[name^="valuations-date"] {
  width: 110px;
}

.valuations-input table input[name^="valuations-amount"] {
  width: 120px;
}

.valuations-input table input[name^="valuations-note"] {
  width: 300px;
}

.valuations-input table td {
  padding: 0;
}

.valuations-input table .control-group {
  margin: 0;
}
