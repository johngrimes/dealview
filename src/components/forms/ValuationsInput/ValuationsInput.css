.add-valuation-button {
  margin-top: 8px;
}

.valuations-input table .valuations-amount input,
.valuations-input table .valuations-note input,
.valuations-input table .react-datepicker__input-container input {
  border: 0;
  box-shadow: none;
  width: 100px;
  padding: 8px 16px;
}

.valuations-input table input[name^="valuations-date"] {
  width: 116px;
}

.valuations-input table input[name^="valuations-amount"] {
  width: 120px;
}

.valuations-input table input[name^="valuations-note"] {
  width: 300px;
}

.valuations-input table td {
  padding: 0;
}

.valuations-input table .control-group {
  margin: 0;
}

.valuations-input .errors {
  width: 540px !important;
}

.valuations-input .delete-valuation-button {
  margin: 0 8px;
  background: transparent url(./delete.svg) no-repeat top left;
  background-size: 18px;
  width: 18px;
  height: 18px;
  border: 0;
}
