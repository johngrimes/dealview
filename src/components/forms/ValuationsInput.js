// @flow

import React from 'react'
import DatePicker from 'react-datepicker'
import moment from 'moment'

import InputField from '../../components/forms/InputField.js'

import './ValuationsInput.css'
import 'react-datepicker/dist/react-datepicker-cssmodules.css'

type Valuation = {
  date: string,
  amount: number,
  note: string
}

export type Valuations = Array<Valuation>

export const ValuationsEmpty = []

type Props = {
  name: string,
  value: Valuations,
  notifyChange?: (valuations: Valuations) => void
}

type State = {
  value: Valuations
}

class ValuationsInput extends React.Component {
  props: Props
  state: State
  handleChange: (i: number, field: string, value: string|number) => void
  handleAddValuation: (event: Event) => true

  constructor(props: Props) {
    super(props)
    this.state = { value: this.props.value }

    this.handleChange = (i, field, value) => {
      const newState = this.state
      newState.value[i][field] = value
      this.setState(newState)
      if (this.props.notifyChange) this.props.notifyChange(newState.value)
    }

    this.handleAddValuation = (event) => {
      const newState = this.state
      newState.value.push({ date: '1983-06-21', amount: 1, note: 'Some note' })
      this.setState(newState)
      if (this.props.notifyChange) this.props.notifyChange(newState.value)
      return true
    }
  }

  render() {
    const valuations = this.state.value.map((v, i) => {
      return <tr key={i}>
        <td><DatePicker name={`valuations-date-${i}`} dateFormat='YYYY-MM-DD' selected={moment(v.date, 'YYYY-MM-DD')}
          onChange={(value) => this.handleChange(i, 'date', value.format('YYYY-MM-DD'))} /></td>
        <td><InputField name={`valuations-amount-${i}`} type='number' value={v.amount.toString()}
          notifyChange={(value) => this.handleChange(i, 'amount', parseInt(value, 10))} /></td>
        <td><InputField name={`valuations-note-${i}`} type='text' value={v.note}
          notifyChange={(value) => this.handleChange(i, 'note', value)} /></td>
      </tr>
    })
    const header = valuations.length > 0 ? <thead>
      <tr>
        <th>Date</th>
        <th>Amount</th>
        <th>Note</th>
      </tr>
    </thead> : null
    return (
      <div className='valuations-input control-group'>
        <table className='table'>
          {header}
          <tbody>{valuations}</tbody>
        </table>
        <button className='button add-valuation-button' type='button' onClick={this.handleAddValuation}>+&nbsp;&nbsp;Add</button>
      </div>
    )
  }
}

export default ValuationsInput
