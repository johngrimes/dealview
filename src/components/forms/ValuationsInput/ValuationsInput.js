// @flow

import React from 'react'
import DatePicker from 'react-datepicker'
import moment from 'moment'
import _ from 'lodash'
import type { Moment } from 'moment'

import { DateDisplayFormat, DateStorageFormat } from 'types/commonTypes'
import { ValuationDefault, compareValuationsByDate } from 'types/valuations'
import HiddenField from 'components/forms/HiddenField'
import * as Validations from 'utils/FormValidation'
import type { Valuations } from 'types/valuations'
import type { FieldErrors } from 'utils/FormValidation'

import 'components/forms/ValuationsInput/ValuationsInput.css'
import 'react-datepicker/dist/react-datepicker-cssmodules.css'

type Props = {
  +name: string,
  +valuations?: Valuations,
  +minDate?: Moment,
  +maxDate?: Moment,
  +errors?: FieldErrors,
  +forceErrorDisplay?: boolean,
  +focus?: string,
  +onChange?: (valuations: Valuations) => void,
  +onFocus?: (fieldName: string) => void
}

type State = {
  +valuations?: Valuations,
  +touched: boolean,
}

class ValuationsInput extends React.Component {
  props: Props
  state: State
  _refs: { [fieldName: string]: { focus: () => void } }
  handleChange: (i: number, field: string, value: string|number) => void
  handleAddValuation: () => void
  handleDeleteValuation: (index: number) => void
  handleFocus: (event: { target: { name: string } }) => void
  handleBlur: (event: { relatedTarget: { name: string } | null }) => void
  setFocus: () => void

  constructor(props: Props) {
    super(props)
    const valuations = this.props.valuations
      ? this.props.valuations.sort(compareValuationsByDate)
      : this.props.valuations
    this.state = {
      valuations,
      touched: false,
    }
    this._refs = {}

    this.handleChange = this.handleChange.bind(this)
    this.handleAddValuation = this.handleAddValuation.bind(this)
    this.handleDeleteValuation = this.handleDeleteValuation.bind(this)
    this.handleFocus = this.handleFocus.bind(this)
    this.handleBlur = this.handleBlur.bind(this)
    this.setFocus = this.setFocus.bind(this)
  }

  handleChange(i: number, field: string, value: string|number): void {
    if (this.state.valuations === undefined) return
    this.setState(
      prevState => {
        const updatedValuations = prevState.valuations.slice()
        updatedValuations[i] = { ...updatedValuations[i], [field]: value }
        return {
          ...prevState,
          valuations: updatedValuations,
          touched: true,
        }
      },
      () => {
        if (this.state.valuations && this.props.onChange) {
          this.props.onChange(this.state.valuations)
        }
      }
    )
  }

  handleAddValuation(): void {
    const updatedValuations = this.state.valuations || []
    updatedValuations.push(ValuationDefault)
    this.setState(
      () => ({ valuations: updatedValuations }),
      () => { if (this.props.onChange) this.props.onChange(updatedValuations) }
    )
  }

  handleDeleteValuation(index: number): void {
    this.setState(
      prevState => ({ valuations: prevState.valuations.filter((_, i) => i !== index) }),
      () => { if (this.props.onChange && this.state.valuations) this.props.onChange(this.state.valuations) }
    )
  }

  handleFocus(event: { target: { name: string } }): void {
    const target = event.target
    if (this.props.onFocus) {
      this.props.onFocus(target.name)
    }
  }

  setFocus(): void {
    if (this.props.focus && this._refs[this.props.focus]) {
      this._refs[this.props.focus].focus()
    }
  }

  handleBlur(event: { relatedTarget: { name: string } | null }): void {
    if (event.relatedTarget &&
        !event.relatedTarget.name.match(`^${this.props.name}`)) {
      this.setState(prevState => ({
        valuations: prevState.valuations.sort(compareValuationsByDate),
      }))
    }
  }

  componentWillReceiveProps(props: Props) {
    if (props.valuations) {
      const sortedValuations = props.valuations.sort(compareValuationsByDate)
      this.setState(() => ({ valuations: sortedValuations }))
    }
  }

  componentDidMount() { this.setFocus() }

  componentDidUpdate() { this.setFocus() }

  render() {
    const { errors, forceErrorDisplay, minDate, maxDate } = this.props
    const { touched } = this.state
    const valuations = this.state.valuations === undefined
      ? []
      : this.state.valuations.map((v, i) => {
        return <tr key={i} className={v.type && v.type !== 'none'
            ? `valuation valuation-${v.type}`
            : 'valuation'}>
          <td className='valuations-date'>
            <DatePicker name={`valuations-date-${i}`} dateFormat={DateDisplayFormat}
              selected={v.date ? moment(v.date, DateStorageFormat) : undefined}
              showYearDropdown
              minDate={minDate} maxDate={maxDate}
              autoFocus={this.props.focus === `valuations-date-${i}`}
              onChange={(moment) => this.handleChange(i, 'date', moment.format(DateStorageFormat))}
              onFocus={this.handleFocus} onBlur={this.handleBlur} />
          </td>
          <td className='valuations-amount'>
            <input name={`valuations-amount-${i}`} type='number' min='1'
              value={v.amount ? v.amount.toString() : ''} placeholder='Value'
              onChange={(event) => this.handleChange(i, 'amount', parseInt(event.target.value, 10))}
              ref={input => this._refs[`valuations-amount-${i}`] = input}
              onFocus={this.handleFocus} onBlur={this.handleBlur} />
          </td>
          <td className='valuations-note'>
            <input name={`valuations-note-${i}`} type='text'
              value={v.note} placeholder='Note'
              onChange={(event) => this.handleChange(i, 'note', event.target.value)}
              ref={input => this._refs[`valuations-note-${i}`] = input}
              onFocus={this.handleFocus} onBlur={this.handleBlur} />
            <HiddenField name={`valuations-type-${i}`} type='hidden' value={v.type} />
          </td>
          <td className='valuations-delete'>
            <button className='delete-valuation-button' type='button' onClick={() => this.handleDeleteValuation(i)} />
          </td>
        </tr>
      })
    const errorTags = []
    if (errors) {
      errors.forEach((msg, i) =>
        errorTags.push(<div key={i} className='error'>{msg}</div>)
      )
    }

    return (
      <div className='valuations-input control-group'>
        <table className='valuations table'>
          <tbody>{valuations}</tbody>
        </table>
        {(forceErrorDisplay || touched) &&
          errorTags.length > 0 && <div className='errors'>{errorTags}</div>}
        <button className='button add-valuation-button' type='button' onClick={this.handleAddValuation}>+&nbsp;&nbsp;Add</button>
      </div>
    )
  }

  static validate(valuations: Valuations): FieldErrors {
    return []
      .concat(ValuationsInput.valuationsDontShareDates(valuations))
  }

  static valuationsDontShareDates(valuations: Valuations): FieldErrors {
    return Validations.validate(
      valuations,
      vals => vals.length === _.uniq(vals.map(v => v.date)).length,
      'Valuations must all be on different dates'
    )
  }
}

export default ValuationsInput
