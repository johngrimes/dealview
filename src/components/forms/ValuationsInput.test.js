/* global expect */

import ValuationsInput from 'components/forms/ValuationsInput.js'

import React from 'react'
import { shallow, mount } from 'enzyme'
import DatePicker from 'react-datepicker'

describe('ValuationsInput', () => {
  const valuations = [
    {
      date: '1983-06-21',
      amount: 18700000,
      note: 'Initial purchase price',
    },
    {
      date: '1992-03-15',
      amount: 36000000,
      note: 'Bank valuation - refinance',
    },
    {
      date: '2001-12-25',
      amount: 75000000,
      note: 'Sworn valuation',
    },
  ]

  it('should render', () => {
    const props = {
      name: 'value',
      valuations,
    }
    const wrapper = shallow(<ValuationsInput {...props} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('should be tolerant of an undefined valuations prop', () => {
    const props = {
      name: 'value',
    }
    const wrapper = shallow(<ValuationsInput {...props} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('should call onChange when an input changes', () => {
    const props = {
      name: 'value',
      valuations,
      onChange: jest.fn(),
    }
    const wrapper = shallow(<ValuationsInput {...props} />)
    wrapper.find('input[name="valuations-note-0"]').prop('onChange')({ target: { value: 'Initial robbery' } })
    expect(props.onChange).toHaveBeenCalledWith(valuations.map((val, i) => {
      return (i === 0)
        ? { ...val, note: 'Initial robbery' }
        : val
    }))
  })

  it('should call onFocus when an input is focused', () => {
    const props = {
      name: 'value',
      valuations,
      onFocus: jest.fn(),
    }
    const wrapper = shallow(<ValuationsInput {...props} />)
    wrapper.find('input[name="valuations-note-1"]').prop('onFocus')({ target: { name: 'valuations-note-1' } })
    expect(props.onFocus).toHaveBeenCalledWith('valuations-note-1')
  })

  it('should focus a date input when the focus prop is passed', () => {
    const props = {
      name: 'value',
      valuations,
      focus: 'valuations-date-0',
    }
    const wrapper = shallow(<ValuationsInput {...props} />)
    expect(wrapper.find(DatePicker).first().prop('autoFocus')).toBe(true)
  })

  const fields = ['amount', 'note']
  fields.forEach(field => {
    it(`should focus an ${field} input when the focus prop is passed`, () => {
      const props = {
        name: 'value',
        valuations,
      }
      const wrapper = mount(<ValuationsInput {...props} />)
      const spy = jest.spyOn(
        wrapper.find(`input[name="valuations-${field}-0"]`).getNode(),
        'focus'
      )
      wrapper.setProps({ focus: `valuations-${field}-0` })
      expect(spy).toHaveBeenCalled()
    })
  })
})
