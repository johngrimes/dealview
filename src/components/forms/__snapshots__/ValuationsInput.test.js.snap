// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ValuationsInput should be tolerant of an undefined valuations prop 1`] = `
<div
  className="valuations-input control-group"
>
  <table
    className="table"
  >
    <tbody />
  </table>
  <button
    className="button add-valuation-button"
    onClick={[Function]}
    type="button"
  >
    +  Add
  </button>
</div>
`;

exports[`ValuationsInput should render 1`] = `
<div
  className="valuations-input control-group"
>
  <table
    className="table"
  >
    <tbody>
      <tr>
        <td
          className="valuations-date"
        >
          <t
            autoFocus={false}
            dateFormat="YYYY-MM-DD"
            dateFormatCalendar="MMMM YYYY"
            disabled={false}
            disabledKeyboardNavigation={false}
            dropdownMode="scroll"
            monthsShown={1}
            name="valuations-date-0"
            onBlur={[Function]}
            onChange={[Function]}
            onClickOutside={[Function]}
            onFocus={[Function]}
            onMonthChange={[Function]}
            onSelect={[Function]}
            popoverAttachment="top left"
            popoverTargetAttachment="bottom left"
            popoverTargetOffset="10px 0"
            selected={"1983-06-20T14:00:00.000Z"}
            showYearDropdown={true}
            tetherConstraints={
              Array [
                Object {
                  "attachment": "together",
                  "to": "window",
                },
              ]
            }
            utcOffset={600}
            withPortal={false}
          />
        </td>
        <td
          className="valuations-amount"
        >
          <input
            name="valuations-amount-0"
            onChange={[Function]}
            onFocus={[Function]}
            placeholder="Value"
            type="number"
            value="18700000"
          />
        </td>
        <td
          className="valuations-note"
        >
          <input
            name="valuations-note-0"
            onChange={[Function]}
            onFocus={[Function]}
            placeholder="Note"
            type="text"
            value="Initial purchase price"
          />
        </td>
      </tr>
      <tr>
        <td
          className="valuations-date"
        >
          <t
            autoFocus={false}
            dateFormat="YYYY-MM-DD"
            dateFormatCalendar="MMMM YYYY"
            disabled={false}
            disabledKeyboardNavigation={false}
            dropdownMode="scroll"
            monthsShown={1}
            name="valuations-date-1"
            onBlur={[Function]}
            onChange={[Function]}
            onClickOutside={[Function]}
            onFocus={[Function]}
            onMonthChange={[Function]}
            onSelect={[Function]}
            popoverAttachment="top left"
            popoverTargetAttachment="bottom left"
            popoverTargetOffset="10px 0"
            selected={"1992-03-14T14:00:00.000Z"}
            showYearDropdown={true}
            tetherConstraints={
              Array [
                Object {
                  "attachment": "together",
                  "to": "window",
                },
              ]
            }
            utcOffset={600}
            withPortal={false}
          />
        </td>
        <td
          className="valuations-amount"
        >
          <input
            name="valuations-amount-1"
            onChange={[Function]}
            onFocus={[Function]}
            placeholder="Value"
            type="number"
            value="36000000"
          />
        </td>
        <td
          className="valuations-note"
        >
          <input
            name="valuations-note-1"
            onChange={[Function]}
            onFocus={[Function]}
            placeholder="Note"
            type="text"
            value="Bank valuation - refinance"
          />
        </td>
      </tr>
      <tr>
        <td
          className="valuations-date"
        >
          <t
            autoFocus={false}
            dateFormat="YYYY-MM-DD"
            dateFormatCalendar="MMMM YYYY"
            disabled={false}
            disabledKeyboardNavigation={false}
            dropdownMode="scroll"
            monthsShown={1}
            name="valuations-date-2"
            onBlur={[Function]}
            onChange={[Function]}
            onClickOutside={[Function]}
            onFocus={[Function]}
            onMonthChange={[Function]}
            onSelect={[Function]}
            popoverAttachment="top left"
            popoverTargetAttachment="bottom left"
            popoverTargetOffset="10px 0"
            selected={"2001-12-24T14:00:00.000Z"}
            showYearDropdown={true}
            tetherConstraints={
              Array [
                Object {
                  "attachment": "together",
                  "to": "window",
                },
              ]
            }
            utcOffset={600}
            withPortal={false}
          />
        </td>
        <td
          className="valuations-amount"
        >
          <input
            name="valuations-amount-2"
            onChange={[Function]}
            onFocus={[Function]}
            placeholder="Value"
            type="number"
            value="75000000"
          />
        </td>
        <td
          className="valuations-note"
        >
          <input
            name="valuations-note-2"
            onChange={[Function]}
            onFocus={[Function]}
            placeholder="Note"
            type="text"
            value="Sworn valuation"
          />
        </td>
      </tr>
    </tbody>
  </table>
  <button
    className="button add-valuation-button"
    onClick={[Function]}
    type="button"
  >
    +  Add
  </button>
</div>
`;
